src/rulesets/LinearAlgebra/dense.jl

```diff
@@ -6,13 +6,25 @@ function frule((_, Δx, Δy), ::typeof(dot), x, y)
     return dot(x, y), dot(Δx, y) + dot(x, Δy)
 end
 
-function rrule(::typeof(dot), x, y)
+function rrule(::typeof(dot), x::AbstractArray, y::AbstractArray)
```

Comment:
Is it a problem that we have now lost e.g. `dot(1+1im, 2+2im)` ?
I guess if we didn't have tests then it doesn't.

If we are restricting this, do we know current one works correctly on arrays of arrays?
the primal `dot` does.

Reply:
Oh, good point about the scalars. Maybe that should be a separate scalar rule. 

More mixed up things, it's a bit of a mess really. `dot([1],2)` works, so does `dot([1,[2,3]], (1,(2,3)))`. But `dot([1,2,3], [1,[2,3]])` does not. I presume that sometimes with arbitrary iterators, AD could just iterate them, and that might be better than turning them into arrays.

Reply:
Before:
```
julia> gradient(dot, 1, 2)
(2, 1)

julia> gradient(dot, [[1],[2]], [[3],[4]])
([[3], [4]], [[1], [2]])

julia> gradient(dot, [1,[2]], [3,[4]])
(Any[3, [4]], Any[1, [2]])

julia> gradient(dot, [1,[2]], (3,4))
((3, 4), Any[1, [2]])

julia> gradient(dot, [1,[2]], (3,(4,)))
ERROR: MethodError: no method matching *(::Tuple{Int64}, ::Int64)
```
After:
```
julia> gradient(dot, 1, 2)
(2, 1)

julia> gradient(dot, [[1],[2]], [[3],[4]])
([[3], [4]], [[1], [2]])

julia> gradient(dot, [1,[2]], [3,[4]])
(Any[3, [4]], Any[1, [2]])

julia> gradient(dot, [1,[2]], (3,4))
(Any[3, [4]], (1, 2))

julia> gradient(dot, [1,[2]], (3,(4,)))
(Any[3, [4]], (1, (2,)))
```